hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieCreateRecordUtils.scala

```diff
@@ -0,0 +1,280 @@
+    }
+    true
+  }
+
+  private def validateMetaFieldsInAvroRecords(avroRec: GenericRecord): Unit = {
+    val avroSchema = avroRec.getSchema;
+    if (avroSchema.getField(HoodieRecord.RECORD_KEY_METADATA_FIELD) == null
+      || avroSchema.getField(HoodieRecord.COMMIT_TIME_METADATA_FIELD) == null
+      || avroSchema.getField(HoodieRecord.COMMIT_SEQNO_METADATA_FIELD) == null
+      || avroSchema.getField(HoodieRecord.PARTITION_PATH_METADATA_FIELD) == null
+      || avroSchema.getField(HoodieRecord.FILENAME_METADATA_FIELD) == null) {
+      throw new HoodieException("Metadata fields missing from avro prepared record.")
+    }
+  }
+
+  def getHoodieKeyAndMaybeLocationFromAvroRecord(keyGenerator: Option[BaseKeyGenerator], avroRec: GenericRecord,
+                                                 isPrepped: Boolean): (HoodieKey, Option[HoodieRecordLocation]) = {
+    val recordKey = if (isPrepped) {
+      avroRec.get(HoodieRecord.RECORD_KEY_METADATA_FIELD).toString
+    } else {
+      keyGenerator.get.getRecordKey(avroRec)
+    };
+
+    val partitionPath = if (isPrepped) {
+      avroRec.get(HoodieRecord.PARTITION_PATH_METADATA_FIELD).toString
+    } else {
+      keyGenerator.get.getPartitionPath(avroRec)
+    };
+
+    val hoodieKey = new HoodieKey(recordKey, partitionPath)
+    val instantTime: Option[String] = if (isPrepped) {
+      Option(avroRec.get(HoodieRecord.COMMIT_TIME_METADATA_FIELD)).map(_.toString)
+    }
+    else {
+      None
+    }
+    val fileName: Option[String] = if (isPrepped) {
+      Option(avroRec.get(HoodieRecord.FILENAME_METADATA_FIELD)).map(_.toString)
+    }
+    else {
+      None
+    }
+    val recordLocation: Option[HoodieRecordLocation] = if (instantTime.isDefined && fileName.isDefined) {
+      val fileId = FSUtils.getFileId(fileName.get)
+      Some(new HoodieRecordLocation(instantTime.get, fileId))
+    } else {
+      None
+    }
+    (hoodieKey, recordLocation)
+  }
+
+  def getHoodieKeyAndMayBeLocationFromSparkRecord(sparkKeyGenerator: Option[SparkKeyGeneratorInterface],
+                                                  sourceRow: InternalRow, schema: StructType,
+                                                  isPrepped: Boolean): (HoodieKey, Option[HoodieRecordLocation]) = {
+    val recordKey = if (isPrepped) {
+      sourceRow.getString(HOODIE_META_COLUMNS_NAME_TO_POS.get(HoodieRecord.RECORD_KEY_METADATA_FIELD));
+    } else {
+      sparkKeyGenerator.get.getRecordKey(sourceRow, schema).toString
+    }
+
+    val partitionPath = if (isPrepped) {
+      sourceRow.getString(HOODIE_META_COLUMNS_NAME_TO_POS.get(HoodieRecord.PARTITION_PATH_METADATA_FIELD))
+    } else {
+      sparkKeyGenerator.get.getPartitionPath(sourceRow, schema).toString
+    };
+
+    val instantTime: Option[String] = if (isPrepped) {
+      Option(sourceRow.getString(HOODIE_META_COLUMNS_NAME_TO_POS.get(HoodieRecord.COMMIT_TIME_METADATA_FIELD)))
```

Comment:
In tagLocation, it will use fileName to parse instantTime, so every record in this file will use the latest instantTime, there is a little difference in implementation. Does this cause the results inconsistency in some sence?

Reply:
And for mor, tag will still use basefile name to parse. I'm not sure how the latter logic works with this location's timestamp, just raising a question

Reply:
good point. checking the logic to see where exactly we use that instant time down the line.

Reply:
its better to keep the behavior consistent. 
we can use FSUtils.getCommitTime(String fileName) to fetch instant time and use that instead of commit time meta field. 
this will be in line w/ how Simple And Bloom index works as of today.

Reply:
I don't see major usages of this instant time in RecordLocation. but better to keep the behavior consistent.

Reply:
Tried using `FSUtils.getCommitTime(String fileName)`, but it leads to `ArrayIndexOutOfBoundsException`

`java.lang.ArrayIndexOutOfBoundsException: 2
	at org.apache.hudi.common.fs.FSUtils.getCommitTime(FSUtils.java:197) ~[classes/:?]
	at org.apache.hudi.HoodieCreateRecordUtils$.getHoodieKeyAndMaybeLocationFromAvroRecord(HoodieCreateRecordUtils.scala:231) ~[classes/:?]`

I have opened a ticket HUDI-6414 to resolve this separate from this patch. Once HUDI-6414 is fixed, we will update the code in getHoodieKeyAndMaybeLocationFrom[Avro|Spark]Record to get commit time.